### DISCUSSION TITLE
Best way to model Heat Conduction at an interface/thermal contact (SideSetHeatTransfer/GapHeatTransfer/Any better method)?

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#25215] [Q&A Modules: General] [answered]
Hi, I'd like to understand the best way to model heat transfer at an interface between two touching blocks using MOOSE.

The two blocks have different material properties and so there is a discontinuity in the temperature between them, therefore conductance (parameter I can supply) needs to be specified at the interface. Important to note that both blocks are touching. I have looked through a couple of options:

1. GapHeatTransfer (https://mooseframework.inl.gov/syntax/ThermalContact/index.html):

In this option/example, it seems that the gap is not explicitly modeled, but you get to define a conductance, which I have. I was thinking this could potentially work for my case, since I'm modeling a "gap" of zero width (no gap). Looking at the syntax below:

```
 [ThermalContact]
  [thermal_contact]
    type = GapHeatTransfer
    variable = temperature
    primary = '101 201'
    secondary = '100 200'
    emissivity_primary = 0
    emissivity_secondary = 0
    gap_conductance = 1.0e9
  []
[]
```

The issue I'm running into here/not sure of is the use of "primary" and "secondary" to define the surfaces. I understand that they'd be there for cases where you actually are modeling a gap, such that the boundaries defined in "primary" and "secondary" would be the faces on either side of the gap. But when there is NO gap as in my case, how could this work? Would it be okay to define the same boundary for both "primary" and "secondary"? I did that and the solution converged, but before I start looking into whether the results are reasonable I want to be sure it's doing what I need it to do.

2. SideSetHeatTransferKernel (https://mooseframework.inl.gov/source/interfacekernels/SideSetHeatTransferKernel.html)

In this case, the documentation is a little more robust and it seems like it's describing what I want as well. However, I ran into a couple of issues while trying to use it. I implemented:

```
[InterfaceKernels]
  [thermal_contact_conductance]
    type = SideSetHeatTransferKernel
    variable = temp
    neighbor_var = temp
    boundary = boundary_conductance
    conductance = 3e-3
  []
[]
```

And got an error saying that I couldn't use the same variable for either side of the discontinuity. So I edited my code and made two different variables for the two blocks in contact, but that didn't work either.

Then I looked at the example, which changed the variable type to "MONOMIAL" with a comment that it did so to be able to properly model the discontinuity. But when I tried the same, other things broke (e.g, I got an error that I was trying to use nodal BCs with an "elemental" variable. I'm not 100% familiar with the theory behind that).

So I guess given my description, which one more accurately captures what I need, and what is the right syntax for implementation?

**Comment by lindsayad** [comment]
For a conforming mesh (which it sounds like you have), e.g. a mesh where neighboring elements at the interface share the same face, the latter object (the interface kernel) is the way to go. That object does this error check which you are running into:
```c++
  if (_var.number() == _neighbor_var.number() && _var.isNodal())
    mooseError(
        "Variable and neighbor variable are the same, but they are not elemental variables.");
```
This error check is there to ensure you can properly model a discontinuity in temperature. If you used the same Lagrange variable on both sides of the interface, you would be forcing the temperature value to be continuous there.

> So I edited my code and made two different variables for the two blocks in contact, but that didn't work either.

Can you share what error message you got?

> But when I tried the same, other things broke (e.g, I got an error that I was trying to use nodal BCs with an "elemental" variable. I'm not 100% familiar with the theory behind that).

Elemental variables like monomials do not have degrees of freedom on nodes, which is where `NodalBC` objects are applied. It sounds like you are also trying to apply a NodalBC on the interface. I would not do that unless you have a good reason. The kernel you are trying to use should be sufficient on its own to ensure your system of equations have a unique solution.

If you do use a discontinuous finite element family you will need to use DGKernels on the interior of your domains/blocks. I don't know if that's what you want. Probably the easiest thing is two lagrange variables defined on either side of the interface.

It looks like @zachmprince was the original author of this object, so perhaps he has thoughts as well

**Reply by richmondodufisan** [reply]
> If you used the same Lagrange variable on both sides of the interface, you would be forcing the temperature value to be continuous there.

I see, that makes sense. And yes, my mesh is conforming. I updated my input file to have two variables for either side of the interface, and added the interface kernel. Here's the error I got:

```
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'gap_Tbulk', requested by 'thermal_contact_conductance' is not defined on boundary boundary_conductance
Material property 'gap_h_primary', requested by 'thermal_contact_conductance' is not defined on boundary boundary_conductance
Material property 'gap_h_neighbor', requested by 'thermal_contact_conductance' is not defined on boundary boundary_conductance
Material property 'gap_emissivity_eff_primary', requested by 'thermal_contact_conductance' is not defined on boundary boundary_conductance
Material property 'gap_emissivity_eff_neighbor', requested by 'thermal_contact_conductance' is not defined on boundary boundary_conductance
```

which I don't know why I got, because according to the documentation, only `neighbor_var` and `variable` are required parameters. I wanted to just set all the properties to zero, but then there is `gap_Tbulk` which is supposed to be the bulk temperature of the gap. Except here, there's no gap.

**Reply by lindsayad** [reply]
Hmm yea how that kernel has coded up those parameters is a little unfortunate. I assume you only want to model conduction for now, and not radiation and convection? If so, then you can just add:
```
Tbulk_mat = 0
h_primary = 0
h_neighbor = 0
emissivity_eff_primary = 0
emissivity_eff_neighbor = 0
```
to your input file block for that object

**Reply by richmondodufisan** [reply]
So I tried that and the solution isn't converging at all. Material properties and everything are fine, I'm actually trying to replicate a simulation done in COMSOL. This is the COMSOL thermal contact documentation:

https://doc.comsol.com/5.5/doc/com.comsol.help.heat/heat_ug_ht_features.09.074.html

And I agree with you that SideSetHeatTransferKernel seems closest to what I'm looking for, but the solution doesn't converge.

On the other hand, using GapHeatTransfer as I described in the original post, i.e:

```
[ThermalContact]
  [thermal_contact]
    type = GapHeatTransfer
    variable = temp
    primary = 'boundary_conductance'
    secondary = 'boundary_conductance'
    gap_conductance = 3e-5
  []
[]
```

converges. I looked through the values returned and saw that they are somewhat close to what I expect ("somewhat" because they're still a little off, I'll know for sure if they are fine after I figure out how to apply a heat source on a thin layer (here, if you can help https://github.com/idaholab/moose/discussions/25200). Right now I'm testing with an upper layer that's thicker than its supposed to be). 

But, even though it converges, the fact that the values are the same on both sides of the discontinuity means it isn't modeling the conductance correctly, right?

**Reply by lindsayad** [reply]
I would not expect that to work (e.g. provide the correct conduction flux at the interface) as the flux/residual is proportional to the difference between the interface temperatures, and that difference should be zero for a continuous variable.

For the interface kernel input, what does your mesh block look like? Does the `variable` parameter lie on the side of the `boundary` parameter? That is critical.

**Reply by richmondodufisan** [reply]
Here are some snippets from my input file and I'll try to describe them:

So I actually used MOOSE to redefine the blocks:

```
  [sample_block]
    type = SubdomainBoundingBoxGenerator
    input = sample_mesh
    block_id = 1
    top_right = '40 20 0'
    bottom_left = '-40 -20 -40'
  []
  [doping_block]
    type = SubdomainBoundingBoxGenerator
    input = sample_block
    block_id = 2	
    top_right = '40 20 0.9'
    bottom_left = '-40 -20 0'
  []
```

Such that I have two blocks, one on top of the other. Then I define the layer between them also with MOOSE:

```
  [conductance_area]
    type = ParsedGenerateSideset
	input = top_no_pump
	combinatorial_geometry = '(z > 0-1e-8) & (z < 0+1e-8)'
    new_sideset_name = boundary_conductance
  []
```

And my kernels:

```
  [heat_conduction_doping]
    type = ADHeatConduction
    variable = temp_dop
	thermal_conductivity = k_dope
	block = doping_material
  []
  [heat_conduction_sample]
    type = ADHeatConduction
    variable = temp_samp
	thermal_conductivity = k_samp
	block = sample_material
  []
```

I also have kernels for the time dependent part too. But this is just to show how I applied different kernels for the different variables. I did this same "separation" for all the materials, BCs and ICs.

**Reply by lindsayad** [reply]
Do you have a MOOSE-only input you can share?

**Reply by richmondodufisan** [reply]
Sure, can I email it to you?

**Reply by lindsayad** [reply]
It's easier for me if you just share it here. Then I don't have to `scp` things

**Reply by richmondodufisan** [reply]
I attached them both (using SideSetHeatTransfer and GapHeatTransfer) + the mesh file they both use

[files]

the thin layer is thicker than its supposed to be since I haven't figured out the ThinLayerHeatTransfer thing, and for some reason I noticed the Neumann condition (heat source) I applied seems to be cooling instead of heating (probably a sign or something missing somewhere) but other than that everything is as it should. thanks for your help

**Reply by richmondodufisan** [reply]
Hi @lindsayad, I figured out a workaround that didn't involve using different variables. I modeled the thin layer as an unconnected mesh (i.e actually modeled a gap, see image). I then used GapHeatTransfer, since, if you directly specify the conductance, it converts that to Neumann conditions on each face. So the actual gap distance in my mesh is irrelevant. I was able to replicate the results just fine with minimal error (which I assume is due to mesh refinement, and/or the poor aspect ratio of the elements in the thin layer).

It would be nice if GapHeatTransfer could be extended/edited for a case like mine, but this workaround will do for now. It would also be nice if ThinLayerHeatTransfer could include an option to apply a surface heat source (right now only the option for a volumetric source exists). Thanks again for your help!!

![comsol_v_moose](https://github.com/idaholab/moose/assets/52802063/3b9330a5-6bcd-4d7f-ae4d-afa3e57a6951)
![Screenshot 2023-08-19 022048](https://github.com/idaholab/moose/assets/52802063/89970bfa-645e-42c4-9e42-a484c02cebac)

**Reply by lindsayad** [reply]
Oh good!

**Reply by richmondodufisan** [reply, marked as answer]
Wanted to add an update here and say that I ended up figuring out how to use SideSetHeatTransferKernel for the simulation (using different variables, and using "SideSetsBetweenSubdomainsGenerator" instead of "ParsedGenerateSideset" to describe the interface. Did that because the results I got with GapHeatTransfer kept diverging from the benchmark results as time progressed (error increased). I'm assuming that kernel does not work exactly as I described and takes the gap into consideration. 

Using SideSetHeatTransfer coupled with better mesh refinement gave me more accurate results. In case anyone happens to find this thread in the future.
